src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1713,6 +1713,15 @@ private final void putCharsAt(int index, CharSequence s, int off, int end) {
         }
     }
 
+    private void putCharsAt(int index, String s, int off, int end) {
+        if (isLatin1() && s.isLatin1()) {
+            System.arraycopy(s.value(), off, this.value, index, end);
+            return;
+        }
+        inflate();
```

Comment:
Like in `String.getBytes(byte[], int, byte)` I think we could do an `arraycopy` if both are `false` too, just need to carefully adjust the `index` et.c. In fact the only case that can't use an `arraycopy` in the end is when `s.isLatin1()` and the current sb is already inflated (that's what the `StringLatin1.inflate` branch does in `getBytes`). 

I think if you consolidate/merge this with the logic in `String.getBytes(byte[], int, byte)` as suggested you'll end up with a sizeable improvement on non-latin1 cases too.

Reply:
Done

Reply:
Nice! 

Did this help the non-latin1 case in your microbenchmark?

Reply:
Yes, now it's significantly faster on longer strings
```
            (latin)  (length)    Score     Error   Units
insert         true         8   22.970 ±   0.291   ns/op
insert         true        64   37.636 ±   0.613   ns/op
insert         true       128   45.412 ±   0.876   ns/op
insert         true      1024  134.511 ±  16.446   ns/op

insert        false         8   35.429 ±   2.410   ns/op
insert        false        64   60.654 ±   1.673   ns/op
insert        false       128   80.743 ±   4.798   ns/op
insert        false      1024  368.987 ±  19.309   ns/op
```